ldm/modules/lora_manager.py

```diff
@@ -0,0 +1,62 @@
+import traceback
+
+from pathlib import Path
+from ldm.invoke.globals import global_lora_models_dir
+from .kohya_lora_manager import KohyaLoraManager
+
+
+class KohyaLoraContext:
+    def __init__(self, kohya_manager: KohyaLoraManager):
+        self.kohya=kohya_manager
+
+    def __enter__(self):
+        self.kohya.clear_loras()
+
+    def __exit__(self,*exc):
+        self.kohya.clear_loras()
+
+class LoraCondition:
+    name: str
+    weight: float
+
+    def __init__(self, name, weight: float = 1.0, kohya_manager: KohyaLoraManager=None):
+        self.name = name
+        self.weight = weight
+        self.kohya_manager = kohya_manager
+
+    def __call__(self, model):
+        # TODO: make model able to load from huggingface, rather then just local files
+        path = Path(global_lora_models_dir(), self.name)
+        if path.is_dir():
+            if model.load_attn_procs:
+                file = Path(path, "pytorch_lora_weights.bin")
+                if file.is_file():
+                    print(f">> Loading LoRA: {path}")
+                    model.load_attn_procs(path.absolute().as_posix())
+                else:
+                    print(f"   ** Unable to find valid LoRA at: {path}")
+            else:
+                print(f"   ** Invalid Model to load LoRA")
+        elif self.kohya_manager:
+            self.kohya_manager.apply_lora_model(self.name,self.weight)
+        else:
+            print(f"   ** Unable to load LoRA")
```

Comment:
Maybe something like this would simplify adding future types (such as LyCORIS), omitted all logging/etc.:

```python
# using Any here, but should be type of `model`
Stage = Callable[[str, float, Any], bool]

class LoraConditionChain:
    def __init__(self, stages: list[Stage]):
        self.stages = set(stages)

    def make_chain(self, name: str, weight: Optional[float]):
        stages = self.stages

        def call(model):
            for s in stages:
                if s(name, weight or 1.0, model):
                    return
            print(f"   ** Failed to load LoRA {name}")

        return call

class KohyaLoraCondition:
    def __init__(self, manager: KohyaLoraManager):
        self.manager = manager

    def __call__(self, name: str, weight: float, model):
        return self.manager.apply_lora_model(name, weight)  # would require updating KohyaLoraManager.apply_lora_model to return True/False depending on if the lora was found and loaded


class DiffuserLoraCondition:
    def __call__(self, name: str, weight: float, model):
        path = Path(global_lora_models_dir(), name)
        if path.is_dir():
            if model.load_attn_procs:
                file = Path(path, "pytorch_lora_weights.bin")
                if file.is_file():
                    print(f">> Loading LoRA: {path}")
                    model.load_attn_procs(path.absolute().as_posix())
                    return True
        return False
```

Then, inside of `LoraManager`, you can simply create a chain:

```python
class LoraManager:
    def __init__(self, pipe):
        # Kohya class handles lora not generated through diffusers
        self.kohya = KohyaLoraManager(pipe, global_lora_models_dir())
        self.lora_condition = LoraConditionChain([KohyaCondition(self.kohya), DiffuserCondition()])

    def set_loras_conditions(self, lora_weights: list):
        conditions = []
        if len(lora_weights) > 0:
            for lora in lora_weights:
                conditions.append(self.lora_condition.make_chain(lora.model, lora.weight))

        if len(conditions) > 0:
            return conditions

        return None

    def kohya_context(self) -> KohyaLoraContext:
        return KohyaLoraContext(self.kohya)
```

This is also missing Peft, but... unsure that even matters at this point, there are no Peft LoRA in the wild that I have ever seen.

Reply:
This looks promising. I'll take a closer look tomorrow.

For what it's worth, I've only tested this so far with Kohya LoRAs. I have no idea if the Peft and diffusers LoRAs are working.